bin-ranges-diff/src/main/java/uk/gov/pay/java_lambdas/bin_ranges_diff/App.java

```diff
@@ -1,17 +1,101 @@
+    }
+
+    @Override
+    public Candidate handleRequest(Candidate candidate, final Context context) {
+        logger.info("fn: {}, version: {}.", context.getFunctionName(), context.getFunctionVersion());
+        var digests = getShaDigestsFromS3(candidate.s3Key());
+        logger.info("Object SHAs [candidate: {}] [promoted: {}]", digests.left(), digests.right());
+        return digests.left().equals(digests.right()) 
+            ? Candidate.from(candidate, false)
+            : Candidate.from(candidate, true);
+    }
+
+    private Pair<String, String> getShaDigestsFromS3(String candidateKey) {
+
+//        var promotedBinRangesHeadObject = s3Client.headObject(RequestBuilder.headObjectRequest(S3_PROMOTED_BUCKET_NAME, PROMOTED_BIN_RANGES_S3_KEY));
+//        var existingDigest = Optional.ofNullable(promotedBinRangesHeadObject.metadata().get("sha256MessageDigest"));
+//        
+//        if (existingDigest.isPresent()) {
+//            
+//        }
+
+        var getCandidateBinRangesRequest = RequestBuilder.getObjectRequest(S3_STAGING_BUCKET_NAME, candidateKey);
+        var getPromotedBinRangesRequest = RequestBuilder.getObjectRequest(S3_PROMOTED_BUCKET_NAME, PROMOTED_BIN_RANGES_S3_KEY);
+
+        return getSha256MessageDigests(getCandidateBinRangesRequest, getPromotedBinRangesRequest);
+    }
+
+
+    private Pair<String, String> getSha256MessageDigests(GetObjectRequest getCandidateBinRangesRequest, GetObjectRequest getPromotedBinRangesRequest) {
+        List<GetObjectRequest> requests = Arrays.asList(getCandidateBinRangesRequest, getPromotedBinRangesRequest);
+        Map<String, String> objectDigests = new ConcurrentHashMap<>();
+
+        requests.parallelStream().forEach(req -> {
+            try (ResponseInputStream<GetObjectResponse> s3ObjectInputStream = s3Client.getObject(req);
+                 BufferedReader reader = new BufferedReader(new InputStreamReader(s3ObjectInputStream, StandardCharsets.UTF_8))) {
+
+                logger.debug("Creating SHA-256 digest for {}", req.key());
+                MessageDigest messageDigest = MessageDigest.getInstance("SHA-256");
 
-    public static void main(String... args) {
-        logger.info("Lambda starts");
+                String line;
+                boolean firstLine = true; // skip the first line as it includes the date the file was published
 
-        Handler handler = new Handler();
-        handler.sendRequest();
+                while ((line = reader.readLine()) != null) {
+                    if (firstLine) {
+                        logger.debug("Ignored first line: {}", line);
+                        firstLine = false;
+                        continue;
+                    }
+                    messageDigest.update(line.getBytes(StandardCharsets.UTF_8));
+                }
 
-        logger.info("Lambda ends");
+                byte[] shaBytes = messageDigest.digest();
+                String base64Sha = Base64.getEncoder().encodeToString(shaBytes);
```

Comment:
`Base64.getEncoder()` returns a `Base64.Encoder`, which is threadsafe, we should be able to do with a single `private static` one in this class.

Reply:
done